File: examples/cpp/table/agg/src/table_state.cpp
#include "table_state.hpp"
#include <algorithm>
#include <vector>

namespace table {

int TableState::active_player_count() const {
    return static_cast<int>(std::count_if(seats.begin(), seats.end(),
        [](const auto& p) { return !p.second.is_sitting_out; }));
}

TableState TableState::from_event_book(const angzarr::EventBook& event_book) {
    TableState state;
    for (const auto& page : event_book.pages()) {
        apply_event(state, page.event());
    }
    return state;
}

void TableState::apply_event(TableState& state, const google::protobuf::Any& event_any) {
    const std::string& type_url = event_any.type_url();

    if (type_url.find("TableCreated") != std::string::npos) {
        examples::TableCreated event;
        if (event_any.UnpackTo(&event)) {
            state.table_id = "table_" + event.table_name();
            state.table_name = event.table_name();
            state.game_variant = event.game_variant();
            state.small_blind = event.small_blind();
            state.big_blind = event.big_blind();
            state.min_buy_in = event.min_buy_in();
            state.max_buy_in = event.max_buy_in();
            state.max_players = event.max_players();
            state.action_timeout_seconds = event.action_timeout_seconds();
            state.status = "waiting";
        }
    } else if (type_url.find("PlayerJoined") != std::string::npos) {
        examples::PlayerJoined event;
        if (event_any.UnpackTo(&event)) {
            state.seats[event.seat_position()] = SeatState{
                event.seat_position(),
                event.player_root(),
                event.stack(),
                true,
                false
            };
        }
    } else if (type_url.find("PlayerLeft") != std::string::npos) {
        examples::PlayerLeft event;
        if (event_any.UnpackTo(&event)) {
            state.seats.erase(event.seat_position());
        }
    } else if (type_url.find("PlayerSatOut") != std::string::npos) {
        examples::PlayerSatOut event;
        if (event_any.UnpackTo(&event)) {
            for (auto& [pos, seat] : state.seats) {
                if (seat.player_root == event.player_root()) {
                    seat.is_sitting_out = true;
                    break;
                }
            }
        }
    } else if (type_url.find("PlayerSatIn") != std::string::npos) {
        examples::PlayerSatIn event;
        if (event_any.UnpackTo(&event)) {
            for (auto& [pos, seat] : state.seats) {
                if (seat.player_root == event.player_root()) {
                    seat.is_sitting_out = false;
                    break;
                }
            }
        }
    } else if (type_url.find("HandStarted") != std::string::npos) {
        examples::HandStarted event;
        if (event_any.UnpackTo(&event)) {
            state.hand_count = event.hand_number();
            state.current_hand_root = event.hand_root();
            state.dealer_position = event.dealer_position();
            state.status = "in_hand";
        }
    } else if (type_url.find("HandEnded") != std::string::npos) {
        examples::HandEnded event;
        if (event_any.UnpackTo(&event)) {
            state.current_hand_root.clear();
            state.status = "waiting";
            // Apply stack changes
            for (const auto& [player_hex, delta] : event.stack_changes()) {
                for (auto& [pos, seat] : state.seats) {
                    std::string seat_hex;
                    for (unsigned char c : seat.player_root) {
                        char buf[3];
                        snprintf(buf, sizeof(buf), "%02x", c);
                        seat_hex += buf;
                    }
                    if (seat_hex == player_hex) {
                        seat.stack += delta;
                        break;
                    }
                }
            }
        }
    } else if (type_url.find("ChipsAdded") != std::string::npos) {
        examples::ChipsAdded event;
        if (event_any.UnpackTo(&event)) {
            for (auto& [pos, seat] : state.seats) {
                if (seat.player_root == event.player_root()) {
                    seat.stack = event.new_stack();
                    break;
                }
            }
        }
    }
}

const SeatState* TableState::get_seat(int position) const {
    auto it = seats.find(position);
    return it != seats.end() ? &it->second : nullptr;
}

const SeatState* TableState::find_player_seat(const std::string& player_root) const {
    for (const auto& [pos, seat] : seats) {
        if (seat.player_root == player_root) {
            return &seat;
        }
    }
    return nullptr;
}

int TableState::find_available_seat(int preferred) const {
    // preferred > 0 means explicit seat preference
    if (preferred > 0 && preferred < max_players) {
        if (seats.find(preferred) == seats.end()) {
            return preferred;
        }
    }
    for (int pos = 0; pos < max_players; ++pos) {
        if (seats.find(pos) == seats.end()) {
            return pos;
        }
    }
    return -1;
}

int TableState::next_dealer_position() const {
    if (seats.empty()) {
        return 0;
    }
    std::vector<int> positions;
    for (const auto& [pos, seat] : seats) {
        positions.push_back(pos);
    }
    std::sort(positions.begin(), positions.end());

    int current_idx = 0;
    for (size_t i = 0; i < positions.size(); ++i) {
        if (positions[i] == dealer_position) {
            current_idx = static_cast<int>(i);
            break;
        }
    }
    int next_idx = (current_idx + 1) % static_cast<int>(positions.size());
    return positions[next_idx];
}

} // namespace table

File: examples/csharp/Table/Agg/TableState.cs
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Angzarr;
using Angzarr.Client;
using Angzarr.Examples;

namespace Table.Agg;

/// <summary>
/// State of a single seat at the table.
/// </summary>
public class SeatState
{
    public int Position { get; set; }
    public ByteString PlayerRoot { get; set; } = ByteString.Empty;
    public long Stack { get; set; }
    public bool IsActive { get; set; } = true;
    public bool IsSittingOut { get; set; }
}

/// <summary>
/// Table aggregate state.
/// </summary>
public class TableState
{
    public string TableId { get; set; } = "";
    public string TableName { get; set; } = "";
    public GameVariant GameVariant { get; set; } = GameVariant.Unspecified;
    public long SmallBlind { get; set; }
    public long BigBlind { get; set; }
    public long MinBuyIn { get; set; }
    public long MaxBuyIn { get; set; }
    public int MaxPlayers { get; set; } = 9;
    public int ActionTimeoutSeconds { get; set; } = 30;
    public Dictionary<int, SeatState> Seats { get; } = new();
    public int DealerPosition { get; set; }
    public long HandCount { get; set; }
    public ByteString CurrentHandRoot { get; set; } = ByteString.Empty;
    public string Status { get; set; } = "";

    public bool Exists => !string.IsNullOrEmpty(TableId);
    public int PlayerCount => Seats.Count;
    public int ActivePlayerCount => Seats.Values.Count(s => !s.IsSittingOut);
    public bool IsFull => Seats.Count >= MaxPlayers;

    public SeatState? GetSeat(int position) => Seats.GetValueOrDefault(position);

    public SeatState? FindPlayerSeat(ByteString playerRoot)
    {
        return Seats.Values.FirstOrDefault(s => s.PlayerRoot.Equals(playerRoot));
    }

    public int? FindAvailableSeat(int preferred = -1)
    {
        if (preferred > 0 && preferred < MaxPlayers && !Seats.ContainsKey(preferred))
            return preferred;
        for (var pos = 0; pos < MaxPlayers; pos++)
        {
            if (!Seats.ContainsKey(pos))
                return pos;
        }
        return null;
    }

    public int NextDealerPosition()
    {
        if (Seats.Count == 0) return 0;
        var positions = Seats.Keys.OrderBy(p => p).ToList();
        var currentIdx = 0;
        for (var i = 0; i < positions.Count; i++)
        {
            if (positions[i] == DealerPosition)
            {
                currentIdx = i;
                break;
            }
        }
        var nextIdx = (currentIdx + 1) % positions.Count;
        return positions[nextIdx];
    }

    /// <summary>
    /// StateRouter for fluent state reconstruction.
    /// </summary>
    public static readonly StateRouter<TableState> Router = new StateRouter<TableState>()
        .On<TableCreated>((state, evt) =>
        {
            state.TableId = $"table_{evt.TableName}";
            state.TableName = evt.TableName;
            state.GameVariant = evt.GameVariant;
            state.SmallBlind = evt.SmallBlind;
            state.BigBlind = evt.BigBlind;
            state.MinBuyIn = evt.MinBuyIn;
            state.MaxBuyIn = evt.MaxBuyIn;
            state.MaxPlayers = evt.MaxPlayers;
            state.ActionTimeoutSeconds = evt.ActionTimeoutSeconds;
            state.Status = "waiting";
        })
        .On<PlayerJoined>((state, evt) =>
        {
            state.Seats[evt.SeatPosition] = new SeatState
            {
                Position = evt.SeatPosition,
                PlayerRoot = evt.PlayerRoot,
                Stack = evt.Stack
            };
        })
        .On<PlayerLeft>((state, evt) =>
        {
            state.Seats.Remove(evt.SeatPosition);
        })
        .On<PlayerSatOut>((state, evt) =>
        {
            var seat = state.Seats.Values.FirstOrDefault(s => s.PlayerRoot.Equals(evt.PlayerRoot));
            if (seat != null) seat.IsSittingOut = true;
        })
        .On<PlayerSatIn>((state, evt) =>
        {
            var seat = state.Seats.Values.FirstOrDefault(s => s.PlayerRoot.Equals(evt.PlayerRoot));
            if (seat != null) seat.IsSittingOut = false;
        })
        .On<HandStarted>((state, evt) =>
        {
            state.HandCount = evt.HandNumber;
            state.CurrentHandRoot = evt.HandRoot;
            state.DealerPosition = evt.DealerPosition;
            state.Status = "in_hand";
        })
        .On<HandEnded>((state, evt) =>
        {
            state.CurrentHandRoot = ByteString.Empty;
            state.Status = "waiting";
            foreach (var kvp in evt.StackChanges)
            {
                var playerRoot = ByteString.CopyFrom(Convert.FromHexString(kvp.Key));
                var seat = state.Seats.Values.FirstOrDefault(s => s.PlayerRoot.Equals(playerRoot));
                if (seat != null) seat.Stack += kvp.Value;
            }
        })
        .On<ChipsAdded>((state, evt) =>
        {
            var seat = state.Seats.Values.FirstOrDefault(s => s.PlayerRoot.Equals(evt.PlayerRoot));
            if (seat != null) seat.Stack = evt.NewStack;
        });

    /// <summary>
    /// Build state from an EventBook by applying all events.
    /// </summary>
    public static TableState FromEventBook(EventBook eventBook)
    {
        return Router.WithEventBook(eventBook);
    }
}
